modules/rtdModule/index.js

```diff
@@ -159,130 +180,147 @@ export function init(config) {
-  subModules.forEach(sm => {
-    sm.getData(adUnits, onDataReceived.bind(sm));
+export function setBidRequestsData(fn, reqBidsConfigObj) {
+  initSubModules();
+
+  // delay bidding process only if auctionDelay > 0
+  if (!_moduleConfig.auctionDelay || _moduleConfig.auctionDelay < 1) {
+    return exitHook();
+  }
+
+  const relevantSubModules = subModules.filter(sm => typeof sm.getBidRequestData === 'function');
+  const prioritySubModules = relevantSubModules.filter(sm => !!(sm.config && sm.config.waitForIt));
+  let callbacksExpected = prioritySubModules.length;
+  let isDone = false;
+  let waitTimeout;
+
+  if (!relevantSubModules.length) {
+    return exitHook();
+  }
+
+  if (prioritySubModules.length) {
+    waitTimeout = setTimeout(exitHook, _moduleConfig.auctionDelay);
+  }
+
+  relevantSubModules.forEach(sm => {
+    sm.getBidRequestData(reqBidsConfigObj, onGetBidRequestDataCallback.bind(sm), sm.config, _userConsent)
   });
 
-  function onDataReceived(data) {
-    if (processDone) {
-      return
+  if (!prioritySubModules.length) {
+    return exitHook();
+  }
+
+  function onGetBidRequestDataCallback(auctionId) {
+    utils.logInfo('onGetBidRequestDataCallback', auctionId, this);
+    if (isDone) {
+      return;
     }
-    dataReceived[this.name] = data;
-    if (shouldWaitForAllSubModules || (this.config && this.config.waitForIt)) {
-      callbacksExpected--
+    if (this.config && this.config.waitForIt) {
+      callbacksExpected--;
     }
     if (callbacksExpected <= 0) {
-      clearTimeout(dataWaitTimeout);
-      done();
+      return exitHook();
     }
   }
 
-  function done() {
-    processDone = true;
-    callback(dataReceived);
+  function exitHook() {
+    isDone = true;
+    clearTimeout(waitTimeout);
+    fn.call(this, reqBidsConfigObj);
   }
 }
 
 /**
- * run hook after bids request and before callback
- * get data from provider and set key values to primary ad server
- * @param {function} next - next hook function
- * @param {AdUnit[]} adUnits received from auction
+ * loop through configured data providers If the data provider has registered getTargetingData,
+ * call it, providing ad unit codes, consent data and module params
+ * the sub mlodle will return data to set on the ad unit
+ * this function used to place key values on primary ad server per ad unit
+ * @param {Object} auction object received on auction end event
  */
-export function setTargetsAfterRequestBids(next, adUnits) {
-  getProviderData(adUnits, (data) => {
-    if (data && Object.keys(data).length) {
-      const _mergedData = deepMerge(setDataOrderByProvider(subModules, data));
-      if (Object.keys(_mergedData).length) {
-        setDataForPrimaryAdServer(_mergedData);
-      }
-    }
-    next(adUnits);
-  });
-}
+export function getAdUnitTargeting(auction) {
+  const relevantSubModules = subModules.filter(sm => typeof sm.getTargetingData === 'function');
```

Comment:
I don't understand how this is asynchronous -- can a submodule hang us up here? Seems like we need some kind of timeout?

Reply:
it's not, we decided only bid modifications will be asynchronous and will use the auction delay parameter.
for ad unit targeting we assume the provider has enough time to call the endpoint before.
(personally, I don't mind adding another delay here, but I don't think it's necessary)

Reply:
I'll let this one go for now. If we have a problem, we may need to revisit.